### PR TITLE
clone: retry on download while network error

### PR DESCRIPTION
**Pull request by lucklove**
### What problem does this PR solve? 
When clone from the mirror, it some times return the network error and abort.

### What is changed and how it works?
Retry several times while there is network errors:
- unexpected EOF 
- stream error
- server returned 502 Bad Gateway

### Check List 

Tests 

 - Manual test

Related changes

 - Need to cherry-pick to the release branch

Release notes:

```release-note
NONE
```

### CODE REVIEW

File: pkg/repository/mirror.go [hunk 1]
```diff
@@ -248,18 +249,46 @@ func (l *httpMirror) prepareURL(resource string) string {
...
 // Download implements the Mirror interface
 func (l *httpMirror) Download(resource, targetDir string) error {
 	tmpFilePath := filepath.Join(l.tmpDir, resource)
 	dstFilePath := filepath.Join(targetDir, resource)
 	// downloaded file is stored in a temp directory and the temp directory is
 	// deleted at Close(), in this way an interrupted download won't remain
 	// any partial file on the disk
-	r, err := l.download(l.prepareURL(resource), tmpFilePath, 0)
+	var err error
+	_ = utils.Retry(func() error {
```
Comment (lonng) [inline pkg/repository/mirror.go:275]: I think we should limit the retry count and retry duration.

Comment (lucklove) [reply pkg/repository/mirror.go:275]: The utils.Retry has default value for that:
```golang
func Retry(doFunc func() error, opts ...RetryOption) error {
	var cfg RetryOption
	if len(opts) > 0 {
		cfg = opts[0]
	} else {
		cfg = RetryOption{
			Attempts: defaultAttempts,
			Delay:    defaultDelay,
			Timeout:  defaultTimeout,
		}
	}
```

File: pkg/repository/mirror.go [hunk 2]
```diff
@@ -248,18 +249,46 @@ func (l *httpMirror) prepareURL(resource string) string {
...
+		var r io.ReadCloser
+		if err != nil && l.isRetryable(err) {
+			log.Warnf("failed to download %s(%s), retrying...", resource, err.Error())
+		}
+		if r, err = l.download(l.prepareURL(resource), tmpFilePath, 0); err != nil {
```
Comment (lonng) [inline pkg/repository/mirror.go:280]: If `err !=nil && !l.isRetryable(err)`, the `download` will execute also.

File: pkg/repository/mirror.go [hunk 3]
```diff
@@ -248,18 +249,46 @@ func (l *httpMirror) prepareURL(resource string) string {
...
 	dstFilePath := filepath.Join(targetDir, resource)
 	// downloaded file is stored in a temp directory and the temp directory is
 	// deleted at Close(), in this way an interrupted download won't remain
 	// any partial file on the disk
-	r, err := l.download(l.prepareURL(resource), tmpFilePath, 0)
+	var err error
+	_ = utils.Retry(func() error {
+		var r io.ReadCloser
+		if err != nil && l.isRetryable(err) {
+			log.Warnf("failed to download %s(%s), retrying...", resource, err.Error())
```
Comment (lonng) [inline pkg/repository/mirror.go:278]: Why not move this line after L281?

**Review by lonng — COMMENTED** [review]
Rest LGTM

File: pkg/repository/mirror.go (hunk 3 above)
Comment (lucklove) [reply pkg/repository/mirror.go:278]: Because it may reach to the max retry attempts, so we must print log before we really retry instead of after retry.

**Review by lonng — APPROVED** [review]
LGTM

**Pull request merged**